Removed the condition causing issue 10198 and added tests to test_inequalities.py

Curious72 opened the pull request:
closes #10588 
fixes #10198

---
- [x] no space after or before parentheses (in the tests)
- [x] space after commas but not before
- [x] line length less than 72; for something like line 325 where the expression is inside parentheses, you can simply break before the last Abs, lining it up with the Or at the start of the line
- [x] be careful to not introduce new whitespace errors (`bin/test quality` is your friend)

Curious72 committed:
This is a combination of 2 commits.
 The first commit's message is:

	modified:   ../inequalities.py
	modified:   test_inequalities.py

	modified:   ../inequalities.py
	modified:   test_inequalities.py

smichr commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -314,5 +314,10 @@ def test_issue_8974():
 
 
+def test_issue_10198():
+    assert reduce_inequalities(-1+1/abs(1/x-1)<0)== And(-oo < x, x < 0) or And(0 < x, x < 1/2)
```
nits: please write as `reduce_inequalities(-1 + 1/abs(1/x - 1) < 0) == ...` (notice added whitespace). Also, there must be something wrong because the right side should be `Or(And(-oo < x, x < 0), And(0 < x, x < S(1)/2))`. Notice the difference below:

```
>>> Or(And(-oo < x, x < 0), And(0 < x, x < S(1)/2))
Or(And(-oo < x, x < 0), And(0 < x, x < 1/2))

>>> And(-oo < x, x < 0) or And(0 < x, x < S(1)/2)
And(-oo < x, x < 0)
```

Why did the tests pass with the expression you have? Not sure.

Curious72 commented:
@smichr  i have added whitespaces but do not understand why the tests are passing for that result...when i put the right solution on right hand side , it gives assertion error

smichr commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -314,5 +314,10 @@ def test_issue_8974():
 
 
+def test_issue_10198():
+    assert reduce_inequalities(-1 + 1/abs( 1/x - 1 ) < 0) == And( -oo < x, x < 0 ) or And( 0 < x, x < 1/2 )
```
change first test to

```
reduce_inequalities(-1 + 1/abs( 1/x - 1 ) < 0) == Or(And(-oo < x, x < 0), And(S(0) < x, x < S(1)/2))
```

smichr replied:
Please add the following two tests (the 2nd of which indicates why the condition that n be an Integer is present):

```
assert reduce_abs_inequality(-3 + 1/Abs(1 - 1/x), '<', x) == \
    Or(And(-oo < x, x < 0), And(S(0) < x, x < S(3)/4),
    And(S(3)/2 < x, x < oo))
raises(ValueError, 
    lambda: reduce_abs_inequality(-3 + 1/Abs(1 - 1/sqrt(x)), '<', x))
```

smichr commented:
Sorry, this has fallen way off the radar as I've worked on other things. Just make a little change:

```diff
diff --git a/sympy/solvers/inequalities.py b/sympy/solvers/inequalities.py
index c73052a..086f32f 100644
--- a/sympy/solvers/inequalities.py
+++ b/sympy/solvers/inequalities.py
@@ -318,9 +318,9 @@ def _bottom_up_scan(expr):
         elif expr.is_Pow:
             n = expr.exp

-            if not n.is_Integer or n < 0:
+            if not n.is_Integer:
                 raise ValueError(
-                    "only non-negative integer powers are allowed")
+                    "Only integer powers are allowed on Abs.")

             _exprs = _bottom_up_scan(expr.base)
```

Then modify the test as indicated and this will be ready to go.

Curious72 commented:
@smichr Sure i will make those changes !!

smichr commented:
The reason you have to change the test is that you essentially have something like this:

```
assert 1 == 0 or 2 -> assert 2 -> passes
```

You have to use `S` as I indicated for a couple reasons, but the most significant is that without import division in the test file,`1/2` becomes `0` so `And( 0 < x, x < 1/2 )` -> `And(x < 0, x > 0)`.

Curious72 committed:
modified:   sympy/solvers/inequalities.py
	modified:   sympy/solvers/tests/test_inequalities.py

Curious72 committed:
Merge remote-tracking branch 'upstream/master'

smichr commented:
+1 after tests pass. Thanks for making the changes.

Curious72 committed:
modified:   sympy/solvers/tests/test_inequalities.py

smichr commented on sympy/solvers/inequalities.py:
```diff
@@ -291,10 +291,10 @@ def reduce_abs_inequality(expr, rel, gen):
     reduce_abs_inequalities
     """
-    if gen.is_real is False:
+    """if gen.is_real is False:
```
oops - missed this. Please restore this if-block

smichr commented on sympy/solvers/inequalities.py:
```diff
@@ -489,8 +488,6 @@ def valid(x):
 
 
 def _solve_inequality(ie, s):
-    """ A hacky replacement for solve, since the latter only works for
```
Please don't remove this. When it's not hacky it will be removed. In the meanwhile, it's the only docstring we have for the existence of this function.

Curious72 committed:
modified:   sympy/solvers/inequalities.py

Curious72 committed:
modified:   sympy/solvers/inequalities.py

smichr commented:
There is still space on line 320 which didn't show in the diff (unless you select the line). You can run tests locally with `bin/test quality` and that will check your file for you. Since you will be editing again, please address some of the other whitespace issues that I list above.

Also, your branch is named 'master'. It is best to follow the guide [here](https://github.com/sympy/sympy/wiki/Development-workflow), especially paying attention to "Create a Separate Branch". All your work should be started as a new branch based off of the current master from github. But don't change branches now or you will have to open a new PR and I would rather get this one closed out rather than start over.

smichr commented on sympy/solvers/inequalities.py:
```diff
@@ -317,10 +317,9 @@ def _bottom_up_scan(expr):
-                    "only non-negative integer powers are allowed")
+            
+            if not n.is_Integer:
+                raise ValueError("Only Integer Powers are allowed on Abs.")
```
my fault here: change "on" to "with"

Also, please only capitalize Only (start of sentence) and Abs (SymPy function name).

smichr commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -318,6 +318,13 @@ def test_issue_8974():
 
 
+def test_issue_10198():
+    assert reduce_inequalities(-1 + 1/abs( 1/x - 1 ) < 0) == Or(And(-oo < x, x < 0), And(S(0) < x, x < S(1)/2))
```
- [ ] no space after or before partheses
- [ ] space after commas but not before
- [ ] line length less than 72; for something like line 325 where the expression is inside parentheses, you can simply break before the last Abs, lining it up with the Or at the start of the line
- [ ] be careful to not introduce new whitespace errors (`bin/test quality` is your friend)

Curious72 committed:
modified:   sympy/solvers/inequalities.py

smichr commented:
Will you take a moment to address the previous checklist of whitespace issues?

Curious72 commented:
I have removed all the whitespace issues , and tested by running  bin/test quality , with the line length less than 72 , line no. 325 seems to be more than 72 characters , should i split it up and write some part of it in next line?

smichr commented:
> should i split it up

Yes. That and other things are discussed on the issue page a few comments before this one. I will move the info to the OP (opening paragraph) to make it easier to find.

Curious72 committed:
Removed spaces before and after paranthesis in tests
	split line 325 , as it exceeded 72 characters

smichr commented:
It would be better to make all the changes at once and then push the modified branch. I'll stop the tests and let you fix the line lenghts.

Curious72 committed:
Splitted lines wherever line length was more than 72 in tests

smichr commented:
Thanks -- it's in.

akki commented on sympy/solvers/tests/test_inequalities.py:
```diff
@@ -318,6 +318,18 @@ def test_issue_8974():
+        -1 + 1/abs(1/x - 1) < 0) == Or(
+        And(-oo < x, x < 0), And(S(0) < x, x < S(1)/2)
+        )
+    assert reduce_inequalities(abs(1/sqrt(x)) - 1, x) == Eq(x, -1)
```
I guess we are checking for `n.integer`, so this test case should not be required.

akki commented:
Oh it got merged, the moment I was commenting.
